### PR TITLE
NDEV-2309: Implement NEON_REVISION caching

### PR DESCRIPTION
**Pull request by Deniskore**
1. Implementation of function get_neon_revisions, which retrieves all NEON_REVISIONS and corresponding slots.
2. Implementation of RevisionMap which handles ranges of slots for specific NEON_REVISION string.

### CODE REVIEW

File: evm_loader/lib/src/types/tracer_ch_common.rs [hunk 1]
```diff
@@ -129,3 +138,65 @@ pub struct EthSyncing {
...
+
+    // When deploying a program for the first time it is now only available in the next slot (the slot after the one the deployment transaction landed in).
+    // When undeploying / closing a program the change is visible immediately and the very next instruction even within the transaction can not access it anymore.
+    // When redeploying the program becomes temporarily closed immediately and will reopen with the new version in the next slot.
+    pub fn build_ranges(input: Vec<(u64, String)>) -> Vec<(u64, u64, String)> {
+        let mut ranges = Vec::new();
+
+        if input.is_empty() {
+            return ranges;
+        }
```
Comment (stanislav-tkach) [inline evm_loader/lib/src/types/tracer_ch_common.rs]: I think this condition can be safely removed.

File: evm_loader/lib/src/types/tracer_ch_common.rs [hunk 2]
```diff
@@ -129,3 +138,65 @@ pub struct EthSyncing {
     pub current_block: u64,
     pub highest_block: u64,
 }
+
+pub struct RevisionMap {
+    map: BTreeMap<u64, String>,
+    pub last_update: Instant,
```
Comment (stanislav-tkach) [inline evm_loader/lib/src/types/tracer_ch_common.rs:144]: Nitpick: perhaps it is better to remove `pub` here and add a getter method. Personally I don't like mixing public and private methods in the same structure, but it is up to you.

**Review by stanislav-tkach — DISMISSED** [review]
Looking good to me, though there is one thing that I personally find a little bit confusing. `RevisionMap` is constructed by following steps:
1. A database query.
2. `RevisionMap::build_ranges` method.
3. `RevisionMap::new` constructor.

I think it would be better to either return `RevisionMap` from the `get_neon_revisions` function or return "raw results" from this function and place the `build_ranges` function call inside the `RevisionMap::new` constructor.

File: evm_loader/lib/src/types/tracer_ch_common.rs (hunk 1 above)
Comment (Deniskore) [reply evm_loader/lib/src/types/tracer_ch_common.rs]: Yes, you're right.

File: evm_loader/lib/src/types/tracer_ch_common.rs (hunk 2 above)
Comment (Deniskore) [reply evm_loader/lib/src/types/tracer_ch_common.rs:144]: In my opinion, this is not considered bad practice if we don't have a lot of fields in the structure.

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**